src/main/java/se/kth/jabeja/Jabeja.java

```diff
@@ -64,34 +64,72 @@ private void sampleAndSwap(int nodeId) {
     Node partner = null;
     Node nodep = entireGraph.get(nodeId);
 
+    boolean doRandom = false;
+
     if (config.getNodeSelectionPolicy() == NodeSelectionPolicy.HYBRID
             || config.getNodeSelectionPolicy() == NodeSelectionPolicy.LOCAL) {
-      // swap with random neighbors
-      // TODO
+      Integer[] sampledNeighbors = getNeighbors(nodep);
+      Node candidatePartner = findPartner(nodeId, sampledNeighbors);
+      if (candidatePartner == null) {
+        doRandom = true;
+      } else {
+        partner = candidatePartner;
+      }
     }
 
-    if (config.getNodeSelectionPolicy() == NodeSelectionPolicy.HYBRID
-            || config.getNodeSelectionPolicy() == NodeSelectionPolicy.RANDOM) {
-      // if local policy fails then randomly sample the entire graph
-      // TODO
-    }
+    if (doRandom || config.getNodeSelectionPolicy() == NodeSelectionPolicy.RANDOM)
+      partner = findPartner(nodeId, getSample(nodeId));
 
     // swap the colors
-    // TODO
+    if (partner != null) {
+      int temp_color = nodep.getColor();
+      nodep.setColor(partner.getColor());
+      partner.setColor(temp_color);
+      numberOfSwaps ++;
+    }
   }
 
   public Node findPartner(int nodeId, Integer[] nodes){
+    double alpha = config.getAlpha();
 
     Node nodep = entireGraph.get(nodeId);
-
     Node bestPartner = null;
     double highestBenefit = 0;
 
-    // TODO
-
+    // Iterate all nodes and get the one that maximizes
+    // utility if swapped with nodep
+    for (int idq: nodes) {
+      Node nodeq = entireGraph.get(idq);
+      double oldB = benefit(nodep, nodeq, alpha);
+      double newB = benefitSwap(nodep, nodeq, alpha);
+      if (newB * T > oldB || newB > highestBenefit) {
```

Comment:
```suggestion
      if (newB * T > oldB && newB > highestBenefit) {
```
Shouldnt this be and??